### ISSUE TITLE
[BUG/MITIGATION] XRDP/Xorg not starting correctly.

### ISSUE THREAD

**Issue by Hacksawfred3232** [neutrinolabs/xrdp#923]
This i don't think classifies as a pull request. But this is a mitigation to solve the problem concerning XRDP not starting correctly.

When connecting to RDP on a computer running with XRDP, it does not start the Xorg desktop program for the connecting RDP session.
This is present in the log:

```
[20171024-12:35:10] [INFO ] A connection received from ::1 port 51480
[20171024-12:35:11] [INFO ] ++ created session (access granted): username pi, ip 0.0.0.0:35408 - socket: 12
[20171024-12:35:11] [INFO ] starting Xorg session...
[20171024-12:35:11] [DEBUG] Closed socket 9 (AF_INET6 :: port 5912)
[20171024-12:35:11] [DEBUG] Closed socket 9 (AF_INET6 :: port 6012)
[20171024-12:35:11] [DEBUG] Closed socket 9 (AF_INET6 :: port 6212)
[20171024-12:35:11] [DEBUG] Closed socket 7 (AF_INET6 ::1 port 3350)
[20171024-12:35:11] [DEBUG] Closed socket 8 (AF_INET6 ::1 port 3350)
[20171024-12:35:11] [INFO ] Xorg :12 -auth .Xauthority -config xrdp/xorg.conf -noreset -nolisten tcp  
[20171024-12:35:21] [ERROR] X server for display 12 startup timeout
[20171024-12:35:21] [INFO ] starting xrdp-sessvc - xpid=18021 - wmpid=18020
[20171024-12:35:21] [ERROR] X server for display 12 startup timeout
[20171024-12:35:21] [ERROR] another Xserver might already be active on display 12 - see log
[20171024-12:35:21] [DEBUG] aborting connection...
[20171024-12:35:21] [INFO ] ++ terminated session:  username pi, display :12.0, session_pid 18019, ip 0.0.0.0:35408 - socket: 12
```
Trying to run Xorg as the user xrdp (After modifying the passwd file to allow a shell) prompts this error
`/usr/lib/xorg/Xorg.wrap: Only console users are allowed to run the X server`
Normally this would not cause a problem. However on this version of XRDP, this causes the timeout error you see in the log.

MITIGATION: Due to the fact that i am still able to run this as root, i considered this to be a problem concerning who is allowed to run the Xserver on my system.
Referencing to Bug report [#865653](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865653) on the Debian bug tracker, this was solved by changing a line in the Xwrapper.config file located in /etc/X11.
`allowed_users=console` to `allowed_users=anybody`
This allowed the xrdp user to run Xorg successfully. 

Potential FIX: A way to automate this mitigation for new installs would be to modify the package install script to automatically apply this fix when you install XRDP. I do not know of a way to add the xrdp user as a console user therefor allowing it to run the Xorg program without applying the mitigation. Anyone who  knows how to do this, Please comment.

**Comment by Hacksawfred3232** [comment]
This also appears to have a similar case to #911 with the same fix. Could be due to Xorg?

**Comment by moobyfr** [comment]
I'm not sure that in debian, a package (xrdp) can modify a file (Xwrapper.config) owned by another package (Xorg).

**Comment by Hacksawfred3232** [comment]
Well i have seen packages like Wireshark and Postfix includes scripts with there packages that automatically run to configure the program before finalizing changes. You create a script that is ran as root during the configure stage to modify that file. If you could link me to a document that specifies the security in place when a package is being installed, i could potentially make a script that takes advantage of this. 
Also to note: It must gain access to the /etc/X11 folder to drop in its library's. @moobyfr

**Comment by metalefty** [comment]
What's the issue? Are you complaining about Debian? 
As Dominik says in the Debian bug, 

>Yet another user not reading NEWS and/or README.Debian… what's so hard about it?

**Comment by Hacksawfred3232** [comment]
No, I am saying that you can avoid this problem in future by having a script that will automatically run during config when installing xrdp. So I find that not our issue tag uncalled for because you can easily overcome this one problem during installing and reduce the headaches that people will get when they get this problem as well and they don't know what to do.

**Comment by metalefty** [comment]
We can modify the Xwrapper config during installation but we SHOULD NOT do that because the file is out of the scope of xrdp. Users never expect `make install` changes other software's config file. Changing existing files without user's intention is quite bad manners.

As you know, xrdp doesn't change firewall config to open 3389/tcp port. It must be done by the user him/herself. The same is true for Xwrapper config. Do you know why Xwrapper is introduced to many systems? For security. As well as firewall config, users should intentionally change Xwrapper config on their own responsibility.

> they don't know what to do.

Okay,  let's write it also to our README. Then people can easily know what for, what to do.

**Comment by Hacksawfred3232** [comment]
Okay. If thats what you want to do. I will let you handle the commit, cause i don't have any experience with the github pull request system. @metalefty

**Comment by metalefty** [comment]
I keep this open until I add explanation in our README.

**Comment by kaysond** [comment]
I just wanted to comment that as a relatively new linux user, this issue caused me quite a headache until I stumbled upon this.

I have Ubuntu 17.10 server running, and wanted to have an optional xrdp/openbox session so I can do text editing. I read through many guides, and my setup seemed right but I couldn't get it to work.

This was my problem, and making that change to Xwrapper.config fixed it. Thanks @Hacksawfred3232 

@metalefty I really think a README note is not sufficient. I understand not wanting to modify Xwrapper.config, and I think that makes sense, but there needs to be a pretty obvious and easily-found message for the user. At the very least, there should probably be a comment in the default xrdp.ini file next to that session, and probably something in the xrdp.ini man page. Not sure who maintains the ubuntu package.

**Comment by jwmullally** [comment]
Ran into this same issue when getting xrdp + Xorg to run inside a Fedora 27 container under an arbitrary user.
This was enough to make everything work:
`sed -i 's;param=Xorg;param=/usr/libexec/Xorg;' /etc/xrdp/sesman.ini`

----------
FYI under Fedora, this is how the Xorg and Xorg SUID wrapper's look:
```
$ ls -la /usr/bin/Xorg /usr/libexec/Xorg /usr/libexec/Xorg.wrap
-rwxr-xr-x. 1 root root     273 Apr 24  2017 /usr/bin/Xorg
-rwxr-xr-x. 1 root root 2334792 Apr 24  2017 /usr/libexec/Xorg
-rwsr-xr-x. 1 root root   11352 Apr 24  2017 /usr/libexec/Xorg.wrap

$ cat /usr/bin/Xorg 
#!/bin/sh
basedir=/usr/libexec
if [ -x "$basedir"/Xorg.wrap ]; then
	exec "$basedir"/Xorg.wrap "$@"
else
	exec "$basedir"/Xorg "$@"
fi
```

**Comment by jwmullally** [comment]
For xrdp, running Xorg should not need SUID or even to be run as root, as it doesn't need access to the console/hardware and is purely an in-memory virtual desktop. Much effort was spent in Xorg/distros to enable running as non-root!

**Comment by metalefty** [comment]
I promise I'll address this issue but please give me time.

**Comment by metalefty** [comment]
@Natureshadow 
What do you think of using `/usr/lib/xorg/Xorg` instead of `/usr/bin/Xorg` in Debian?
`/usr/lib/xorg/Xorg` can be run as user privilege. Users need not edit `Xwrapper.config`.

This what @jwmullally said makes sense for me.
> For xrdp, running Xorg should not need SUID or even  to be run as root, as it doesn't need access to the console/hardware and  is purely an in-memory virtual desktop.

**Comment by metalefty** [comment]
@kaysond 
> but there needs to be a pretty obvious and easily-found message for the user.

Sure but let me mend it as a long-term solution. It will be done in #599 #909.
As a short-term workaround, I added a comment in `sesman.ini`.  See #974 and what do you think?

**Comment by kaysond** [comment]
@metalefty TBH, I never even looked at sesman.ini, so that might've not helped me, but it seems like a good solution.

What about a comment in the default xrdp.ini next to the xorg session entry?

I think most people would be looking in that file first. I'm not sure if this is a specific problem to xorg or if other libraries like xvnc or whatever also require root. Maybe there could be a comment somewhere saying make sure the library is executable by user

**Comment by jwmullally** [comment]
FWIW, I think if this particular setting is configured in sesman.ini the comments should be there.

To address @kaysond concern in last comment, I think its enough to add a comment in /etc/xrdp/xrdp.ini at the start of the "; Session Types" section to say something like "Local sesman managed sessions are configured in sesman.ini"

**Comment by metalefty** [comment]
@jwmullally @kaysond I've updated the PR. Can you review?

>  I'm not sure if this is a specific problem to xorg or if other libraries like xvnc or whatever also require root

Potentially, other libraries also have the same issue but actually, the issue hardly occurs with other libs. Because Xvnc is intended to run in user privilege and no hack is needed to run in user priv in most cases.

**Comment by kaysond** [comment]
@metalefty i think thats great. thanks!

**Comment by metalefty** [comment]
Thanks. Of course I'll work toward a long-term solution.

**Comment by metalefty** [comment]
Merged #974.
